### PR TITLE
fix: encode path parameters

### PR DESCRIPTION
**Pull request by eshanholtz**
This PR is intended to be a more robust solution to the URL encoding bug found in #557

### CODE REVIEW

File: src/test/java/com/twilio/Assert.java [hunk 1]
```diff
@@ -30,15 +30,18 @@ public static void assertUrlsEqual(final String expected, final URL actual) {
...
+        String[] actualParts = actual.split("[?#]");
 
         assertEquals(expectedParts.length, actualParts.length);
         assertEquals(expectedParts[0], actualParts[0]);
 
-        if (expectedParts.length == 2) {
+        if (expectedParts.length >= 2) {
             assertQueryStringsEqual(expectedParts[1], actualParts[1]);
         }
+        if (expectedParts.length == 3) {
```
Comment (thinkingserious) [inline src/test/java/com/twilio/Assert.java:42]: Why is this case called out separately, when the above code on line 39 checks for all values `>=2`? Should we call both functions if `expectedParts.length == 3`?

File: src/test/java/com/twilio/http/RequestTest.java [hunk 2]
```diff
@@ -196,13 +205,13 @@ public void testRegionAndEdgeInUrl() throws MalformedURLException {
     }
 
     @Test
-    public void testRegionInConstructor() throws MalformedURLException {
-        final Request request = new Request(HttpMethod.GET, Domains.ACCOUNTS.toString(), "/path/to/something.json?foo=12.34", "region");
+    public void testRegionInConstructor() {
```
Comment (thinkingserious) [inline src/test/java/com/twilio/http/RequestTest.java:208]: Should we keep the old test, and also add this one? (testing with and without `#bar` in the URL)

Comment (eshanholtz) [reply src/test/java/com/twilio/http/RequestTest.java:208]: It does not make any meaningful difference to test with and without the fragment on the end, it just adds more code.

File: src/test/java/com/twilio/Assert.java (hunk 1 above)
Comment (eshanholtz) [reply src/test/java/com/twilio/Assert.java:42]: Note the code inside the conditional blocks. If the number of parts is >= 2, then the url contains a query string, which are there compared. If number of parts is 3, then the url also contains a fragment, which then needs to be compared as well.

File: src/main/java/com/twilio/http/Request.java [hunk 3]
```diff
@@ -167,15 +169,14 @@ private String buildURL() {
...
-            } catch (final MalformedURLException e) {
-                throw new ApiException("Bad URL: " + url, e);
+                host = joinIgnoreNull(".", product, targetEdge, targetRegion, domain);
             }
-        }
 
-        return url;
+            return new URI(parsedUrl.getProtocol(), parsedUrl.getUserInfo(), host, parsedUrl.getPort(),
+                    parsedUrl.getPath(), parsedUrl.getQuery(), parsedUrl.getRef()).toURL().toString();
+        } catch (final MalformedURLException | URISyntaxException e) {
```
Comment (childish-sambino) [inline src/main/java/com/twilio/http/Request.java:177]: 🎉

File: src/test/java/com/twilio/http/RequestTest.java (hunk 2 above)
Comment (childish-sambino) [reply src/test/java/com/twilio/http/RequestTest.java:208]: It's a legitimate concern. Like, what if things only work when _both_ query params and fragments are used but fail when it's just one or the other. Not saying it doesn't work in those cases, but someone later may refactor the logic where it does in fact not work properly, but this test would still pass.

But, yes, in general, I'm against more test code that can be reduced. Test code is just like source code: it needs to be maintained.

Comment (eshanholtz) [reply src/test/java/com/twilio/http/RequestTest.java:208]: I agree, it's just in this case we [already have tests that include just query params](https://github.com/twilio/twilio-java/pull/558/files#diff-d249e028337d87ed2f4499346031488fR71). That said, I don't believe we have a test with no query params and just a fragment, but the way it's split, just query params vs just a fragment is treated equivalently.

**Pull request merged**